pkg/v1/utils/volumes.go

```diff
@@ -0,0 +1,152 @@
+func volumeFieldName(k string) string {
+	parts := strings.Split(strings.TrimPrefix(k, volumeKeyPrefix), "/")
+	return parts[len(parts)-1]
+}
+
+// podUID gets pod UID from the pod log path.
+func podUID(s *specs.Spec) (string, error) {
+	sandboxLogDir := s.Annotations[annotations.SandboxLogDir]
+	if sandboxLogDir == "" {
+		return "", errors.New("no sandbox log path annotation")
+	}
+	fields := strings.Split(filepath.Base(sandboxLogDir), "_")
+	switch len(fields) {
+	case 1: // This is the old CRI logging path
+		return fields[0], nil
+	case 3: // This is the new CRI logging path
+		return fields[2], nil
+	}
+	return "", errors.Errorf("unexpected sandbox log path %q", sandboxLogDir)
+}
+
+// isVolumeKey checks whether an annotation key is for volume.
+func isVolumeKey(k string) bool {
+	return strings.HasPrefix(k, volumeKeyPrefix)
+}
+
+// volumeSourceKey constructs the annotation key for volume source.
+func volumeSourceKey(volume string) string {
+	return volumeKeyPrefix + volume + "/source"
+}
+
+// volumePath searches the volume path in the kubelet pod directory.
+func volumePath(volume, uid string) (string, error) {
+	// TODO: Support subpath when gvisor supports pod volume bind mount.
+	volumeSearchPath := fmt.Sprintf("/var/lib/kubelet/pods/%s/volumes/*/%s", uid, volume)
+	dirs, err := filepath.Glob(volumeSearchPath)
+	if err != nil {
+		return "", err
+	}
+	if len(dirs) != 1 {
+		return "", errors.Errorf("unexpected matched volume list %v", dirs)
+	}
+	return dirs[0], nil
+}
+
+// isVolumePath checks whether a string is the volume path.
+func isVolumePath(volume, path string) (bool, error) {
+	// TODO: Support subpath when gvisor supports pod volume bind mount.
+	volumeSearchPath := fmt.Sprintf("/var/lib/kubelet/pods/*/volumes/*/%s", volume)
+	return filepath.Match(volumeSearchPath, path)
+}
+
+// UpdateVolumeAnnotations add necessary OCI annotations for gvisor
+// volume optimization.
+func UpdateVolumeAnnotations(bundle string, s *specs.Spec) error {
+	var (
+		uid string
+		err error
+	)
+	if IsSandbox(s) {
+		uid, err = podUID(s)
+		if err != nil {
+			// Skip if we can't get pod UID, because this doesn't work
+			// for containerd 1.1.
+			logrus.WithError(err).Error("Can't get pod uid")
+			return nil
+		}
+	}
+	done := make(map[string]struct{})
+	for k, v := range s.Annotations {
+		if !isVolumeKey(k) {
+			continue
+		}
+		if volumeFieldName(k) != "type" {
+			continue
+		}
+		if v != "tmpfs" {
+			// Only tmpfs is supported now.
+			continue
+		}
+		volume := volumeName(k)
+		if _, ok := done[volume]; ok {
+			continue
+		}
+		if uid != "" {
+			// This is a sandbox
+			path, err := volumePath(volume, uid)
+			if err != nil {
+				return errors.Wrapf(err, "get volume path for %q", volume)
+			}
+			s.Annotations[volumeSourceKey(volume)] = path
+		} else {
+			// This is a container
+			for i := range s.Mounts {
+				if yes, _ := isVolumePath(volume, s.Mounts[i].Source); yes {
```

Comment:
It's safer to match s.Mounts[i].Source, with the current annotation's source. Otherwise, it may upgrade to tmpfs a mount which doesn't have annotations associated with it (e.g. a mount with mismatching mount options).

Reply:
We don't have the source in shim for container, because there is no way to get pod UID right now.

Reply:
Wouldn't this incorrectly mark an EmptyDir mount with mismatching rw/ro options as gVisor internal tmpfs, instead of a shared volume mounted externally? The former would cause a distinct tmpfs to be mounted inside each containers.

Reply:
That's why I return an error for sandbox if an tmpfs annotation exists, but `source` is not successfully applied. That guaranteed that as long there is an tmpfs annotation, source should be applied on sandbox.

And the volume name is unique for each pod, so this should work. But of course, this is not the best way this should work, we need to find a way to pass down more information to get rid of this hack.

Reply:
Sounds good. Please leave a comment in the code explaining it and maybe a TODO to fix it later.

Reply:
Done

Reply:
Done